core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -973,11 +989,14 @@ private InFlightState tryUpdateState(RecordState newState, boolean incrementDeli
             }
         }
 
-        private InFlightState startStateTransition(RecordState newState, boolean incrementDeliveryCount) {
+        private InFlightState startStateTransition(RecordState newState, boolean incrementDeliveryCount, int maxDeliveryCount) {
             try {
                 rollbackState = new InFlightState(state, deliveryCount);
+                if (newState == RecordState.AVAILABLE && deliveryCount == maxDeliveryCount) {
```

Comment:
I would use `deliveryCount >= maxDeliveryCount` to be a bit more defensive. If we had a bug in which we failed to archive when the count was equal, we'd at least pick it up the next delivery attempt.